burn-tensor/src/tensor/api/int.rs

```diff
@@ -45,6 +45,22 @@ where
     }
 }
 
+impl<B> Tensor<B, 2, Int>
+where
+    B: Backend,
+{
+    /// Create diagonal matrix.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size of the square matrix.
+    pub fn diagonal(size: usize) -> Tensor<B, 2, Int> {
+        let indices = Tensor::arange(0..size).unsqueeze();
+        let ones = Tensor::ones([size]).unsqueeze();
+        Tensor::<B, 2, Int>::zeros([size, size]).scatter(0, indices, ones)
+    }
```

Comment:
Move this function to the `numeric.rs` file, it will be available for both `Int` and `Float` tensors, avoiding casting.